lib/queue.js

```diff
@@ -223,6 +223,91 @@ Queue.prototype.checkHealth = function (cb) {
   return promise;
 };
 
+Queue.prototype._scanForJobs = function(cursor, set, cb) {
+  // Ideally, we would be able to paginate redis sets like we do sorted ranges.
+  // For an MVP, however, we're aiming to replace our usage of Arena for Bull 1.x
+  // and Bull 1.x calls SMEMBERS -- which is more blocking than SSCAN.
+
+  this.client.sscan(
+    this.toKey('failed'),
+    cursor,
+    'COUNT', '100',
+    (err, results) => {
+      const [nextCursor, ids] = results;
+
+      // A given element may be returned multiple times in SSCAN.
+      // So, we use a set to remove duplicates.
+      // https://redis.io/commands/scan#scan-guarantees
+      for (let id of ids) set.add(id);
+
+      if (nextCursor === '0') {
+        return cb(null, set);
+      }
+
+      this._scanForJobs(cb, nextCursor, set);
+    }
+  );
+}
+
+// TODO(randall): Implement (start, end) parameters
+Queue.prototype._getJobs = function (queueType, start, end, cb) {
+  let promise;
+
+  if (this.paused) {
+    promise = Promise.reject(new Error('closed'));
+  } else {
+    const idsPromise = helpers.deferred();
+    this._ready
+      .then(() => {
+        if (this.paused) {
+          throw new Error('closed');
+        }
+      })
+      .then(() => {
```

Comment:
no need to make this a separate `then` - it can just be part of the previous one